[Benedict Chua] iP

benedictchuajj opened the pull request:
# Duke, the Tsundere you wish you never had
> "This is so cringe and I didn't know what I was ~~smoking~~ thinking when I started this" - Benedict Chua :monkey:

Duke frees your mind of having to remember things you need to do. It's
* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use
* and to top it off, it has an **attitude** you probably never wanted

All you need to do is,
1. download it from [here](https://github.com/benedictchuajj/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE!** 

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

## For Developers:

If you want to clone the repository and run it on your CLI instead of using the jar file, you can follow the instructions here.

From the root of the repository:
```
cd src/main/java
```
Compile all the classes in the package:
```
javac -classpath . duke/Duke.java
```
Finally, run the program:
```
java duke/Duke
```

Some random Java code here:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Some random inline code `print("hello world")` here even though it is clearly not Java

---

~~Please don't judge me too hard for this, I'm just having fun cringing at this version of Duke~~ :clown_face:

Commit:
Add Gradle support

benedictchuajj committed:
Add greet, echo, exit functionalities

benedictchuajj committed:
Add add and list functionalities along with TaskList class

benedictchuajj committed:
Add mark as done functionality along with Task class

benedictchuajj committed:
Add support for tracking three types of task.

Task is currently too broad and does not have specific commands.

With commands, users are less prone to inputting the wrong types of
tasks into the list. Also, users will be able to distinguish between
types of tasks.

Let's
* add ToDo, Deadline, Event that extends from Task
* remove original add function and replace it with respective calls
  to the type of task to act as the command to add task

benedictchuajj committed:
Add automated text UI testing

benedictchuajj committed:
Handle errors for invalid inputs when using Duke

benedictchuajj committed:
Update testing for handling errors

benedictchuajj committed:
Add delete task functionality, change TaskList to use ArrayList from array

benedictchuajj committed:
Update commands to be enum

benedictchuajj committed:
Update JavaDocs for all class descriptions

benedictchuajj committed:
Add saving and loading functionalities

benedictchuajj committed:
Update level progress in Duke

benedictchuajj committed:
Add date and time functionalities to tasks

benedictchuajj committed:
Merge branch 'branch-Level-8'

benedictchuajj committed:
Refactor code to abstract closely related code as classes

benedictchuajj committed:
Organise classes into suitable java packages

benedictchuajj committed:
Remove duplicate Java files

benedictchuajj committed:
Add JUnit tests

benedictchuajj committed:
Create a JAR File

benedictchuajj committed:
Update JavaDocs for all classes

benedictchuajj committed:
Update code to comply with the Coding Standard

benedictchuajj committed:
Add functionality to find a task by searching for a keyword

benedictchuajj committed:
Merge branch 'branch-A-CodingStandard'

benedictchuajj committed:
Merge branch 'branch-Level-9'

benedictchuajj committed:
Update credit for reused work in test classes

benedictchuajj committed:
Update naming to improve readability

benedictchuajj committed:
Merge branch 'master' into add-gradle-support

benedictchuajj committed:
Add gradle support

benedictchuajj committed:
Add checkStyle to detect coding style violations

benedictchuajj committed:
Update code to follow checkStyle

benedictchuajj committed:
Add a GUI to Duke using JavaFX

benedictchuajj committed:
Update JavaDocs for DialogBox and Ui

benedictchuajj committed:
Add functionality to Done command to take in multiple indexes

phuccuongngo99 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,29 @@
+        this.index = index;
+    }
+
+    @Override
```
Can put @inheritDocs here, just a nit :grin:

benedictchuajj replied:
Thanks for the suggestion! Did not that existed 😄

phuccuongngo99 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,105 @@
+     *
+     * @return String of the formatted Task.
+     */
+    public abstract String saveAsString();
```
Can consider changing the name of this method to `convertToString()`

ngsengleng replied:
saveAsString seems to be ok here

benedictchuajj replied:
I agree that convertToString might be a better name since I am returning a String and not really saving it (void function). Thanks for the suggestion! 😃

phuccuongngo99 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,88 @@
+     * @param description String of the task description.
+     * @param dueDate String of the given due date and time.
+     */
+    public Deadline(String completed, String description, String dueDate) {
```
Naming `completed` variable to `completeStatus` to match the intention more closely

benedictchuajj replied:
Good catch! I think it might be even better if I make it into an enumeration now that I noticed this again.

phuccuongngo99 reviewed (commented):
Very well written code with no code violation.
Good use of CheckStyle to ensure no code violation.
Just a few nits with variable naming as It's near impossible to find code violations.
Very well done my dude :grin:

ngsengleng commented on src/main/java/duke/util/Parser.java:
```diff
@@ -99,14 +100,47 @@ public static int filterTaskIndex(String command) throws MissingIndexException {
+     * @return the filtered and parsed int that refers to the index of the task.
+     * @throws MissingIndexException when index is missing (including if it contains only white space).
+     */
+    public static int[] filterMultipleTaskIndexes(String command) throws MissingIndexException {
```
Perhaps a more apt function name could be getAllTaskIndexes?

benedictchuajj replied:
Thanks for the suggestion! I think that it probably a better way of naming the function than my original name.

ngsengleng commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -154,12 +154,12 @@ public void printList(String type, String filterCondition) {
         String[] message = {
                 "You completed a task! Maybe you aren't so incompetent after all.",
-                tasksList.get(index - 1).markTaskAsDone()
+                tasks.get(index - 1).markTaskAsDone()
```
perhaps a String.format suffices? or changing the variable name to plural so it does not violate the naming standard?

benedictchuajj replied:
I have already changed the variable name here already but you're right, I definitely can use String.format to make this portion and similar parts more readable. 😃

ngsengleng commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,95 @@
+        // Checks if file exists, creates if it doesn't
+        if (!this.doesFileExists) {
+            File file = new File(FILE_PATH);
+            boolean result;
```
perhaps isFileCreated is more apt?

benedictchuajj replied:
Thanks for the suggestion! I think it might be a better name 👍

ngsengleng commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,58 @@
+ * @author Benedict Chua
+ */
+public class Ui {
+    private static final String INDENTATION = "     ";
+    private static final String LINE_SEPARATOR = "    ____________________________________________________________";
+    private static final String[] GREETING = {
```
Good declaration of final variables!

ngsengleng reviewed (commented):
Overall code quality is pretty good and creative modifications of the Duke bot including being able to read multiple tasks to be deleted. Nice work! 👍

benedictchuajj commented:
Thanks @phuccuongngo99 and @ngsengleng for your feedbacks and suggestions! I appreciate your time in helping me catch the parts that I could improve on. 🙇

benedictchuajj committed:
Improve code based on feedbacks received in PR review

benedictchuajj committed:
Add assertions to document important assumptions

There are a few methods that have assumptions in their body, either
through switch statements, if-else statements or that another method
returns a value that is supposed to be valid.

Having assertions in these methods would document these important
assumptions that has to hold at various points in the code.

Let's insert assertions for,
* internal invariants
* control-flow invariants
* preconditions
* postconditions
* class invariants

benedictchuajj committed:
Refactor code to improve code quality

There are several parts of the codebase that are not adhering
to the practices that W5.4 Code Quality recommends.

Improving the code quality will make the codebase more readable
and understandable. This can also help in reducing the number of
potential bugs and making the codebase self-explanatory.

Let's,
* Refactor the parts to make it improve readability
* Correct the parts that are actually unsafe practices

benedictchuajj committed:
Refactor code to use Stream

benedictchuajj committed:
Merge pull request #1 from benedictchuajj/branch-A-Assertions

Add assertions to document important assumptions

benedictchuajj committed:
Merge branch 'master' into branch-A-CodeQuality

benedictchuajj committed:
Merge pull request #2 from benedictchuajj/branch-A-CodeQuality

Refactor code to improve code quality

benedictchuajj committed:
Merge pull request #3 from benedictchuajj/branch-A-Streams

Refactor code to use Stream

benedictchuajj committed:
Fix bugs caused by assertions

benedictchuajj committed:
Set up CI

Currently tests are not run automatically when pushing or merging.

Having Continuous Integration can help automate the process of
checking for coding style violations and running tests to check
for regressions.

As a step towards automation, let's set up CI to ensure that code
quality persists.

benedictchuajj committed:
Add exception to catch for invalid index arguments in commands

benedictchuajj committed:
Fix imports in Parser

benedictchuajj committed:
Add reminder functionality to Duke

benedictchuajj committed:
Update code to comply with checkstyle

benedictchuajj committed:
Merge pull request #4 from benedictchuajj/branch-B-Reminders

Add reminder functionality to Duke

benedictchuajj committed:
Improve GUI

benedictchuajj committed:
Merge pull request #5 from benedictchuajj/branch-A-BetterGui

Improve GUI

benedictchuajj committed:
Set up product website (sample SS and user guide)

benedictchuajj committed:
Update README.md

benedictchuajj committed:
Update sample screenshot

benedictchuajj committed:
Merge branch 'master' of https://github.com/benedictchuajj/ip

benedictchuajj committed:
Update sample screenshot

benedictchuajj committed:
Update README.MD

benedictchuajj committed:
Set theme jekyll-theme-cayman

benedictchuajj committed:
Update README.md

benedictchuajj committed:
Update variable name to follow code style
